File: WordPress/Classes/ViewRelated/Reader/WPContentSyncHelper.m
#import "WPContentSyncHelper.h"

@interface WPContentSyncHelper()

@property(nonatomic, readwrite) BOOL isSyncing;
@property(nonatomic, readwrite) BOOL isLoadingMore;
@property(nonatomic, readwrite) BOOL hasMoreContent;

@end

@implementation WPContentSyncHelper

#pragma mark - Life Cycle Methods

- (instancetype)init
{
    self = [super init];
    if (self) {
        _hasMoreContent = YES;
    }
    return self;
}


#pragma mark - Public Methods

#pragma mark - Accessors

- (void)setHasMoreContent:(BOOL)hasMoreContent
{
    if (_hasMoreContent == hasMoreContent) {
        return;
    }

    _hasMoreContent = hasMoreContent;

    if (!_hasMoreContent && [self.delegate respondsToSelector:@selector(hasNoMoreContent)]) {
        [self.delegate hasNoMoreContent];
    }
}


#pragma mark - Syncing

- (BOOL)syncContent
{
    return [self syncContentViaUserInteraction:NO];
}

- (BOOL)syncContentViaUserInteraction
{
    return [self syncContentViaUserInteraction:YES];
}

- (BOOL)syncContentViaUserInteraction:(BOOL)userInteraction
{
    if (self.isSyncing) {
        return NO;
    }

    self.isSyncing = YES;
    [self.delegate syncHelper:self syncContentWithUserInteraction:userInteraction success:^(NSUInteger count) {
        self.hasMoreContent = (count > 0);
        [self syncContentEnded];
    } failure:^(NSError *error) {
        [self syncContentEnded];
    }];

    return YES;
}

- (BOOL)syncMoreContent
{
    if (self.isSyncing || !self.hasMoreContent) {
        return NO;
    }

    self.isSyncing = YES;
    self.isLoadingMore = YES;
    [self.delegate syncHelper:self syncMoreWithSuccess:^(NSUInteger count) {
        self.isLoadingMore = NO;
        self.hasMoreContent = (count > 0);
        [self syncContentEnded];
    } failure:^(NSError *error) {
        [self syncContentEnded];
    }];

    return YES;
}


#pragma mark - Private Methods

- (void)syncContentEnded
{
    self.isSyncing = NO;
    self.isLoadingMore = NO;

    if([self.delegate respondsToSelector:@selector(syncContentEnded)]) {
        [self.delegate syncContentEnded];
    }
}

@end

File: WordPress/Classes/ViewRelated/Reader/WPContentSyncHelper.swift
import UIKit


@objc protocol WPContentSyncHelperDelegate: NSObjectProtocol {
    func syncHelper(syncHelper:WPContentSyncHelper, syncContentWithUserInteraction userInteraction: Bool, success: ((hasMore: Bool) -> Void)?, failure: ((error: NSError) -> Void)?)
    func syncHelper(syncHelper:WPContentSyncHelper, syncMoreWithSuccess success: ((hasMore: Bool) -> Void)?, failure: ((error: NSError) -> Void)?)
    optional func syncContentEnded()
    optional func hasNoMoreContent()
}


class WPContentSyncHelper: NSObject {

    weak var delegate: WPContentSyncHelperDelegate?
    var isSyncing:Bool = false
    var isLoadingMore:Bool = false
    var hasMoreContent:Bool = true {
        didSet {
            if hasMoreContent == oldValue {
                return
            }
            if hasMoreContent == false {
                delegate?.hasNoMoreContent?()
            }
        }
    }


    // MARK: - Syncing

    func syncContent() -> Bool {
        return syncContentWithUserInteraction(false)
    }


    func syncContentWithUserInteraction() -> Bool {
        return syncContentWithUserInteraction(true)
    }


    func syncContentWithUserInteraction(userInteraction:Bool) -> Bool {
        if isSyncing {
            return false
        }

        isSyncing = true

        delegate?.syncHelper(self, syncContentWithUserInteraction: userInteraction, success: {
            [weak self] (hasMore: Bool) -> Void in
            if let weakSelf = self {
                weakSelf.hasMoreContent = hasMore
                weakSelf.syncContentEnded()
            }
        }, failure: {
            [weak self] (error: NSError) -> Void in
            if let weakSelf = self {
                weakSelf.syncContentEnded()
            }
        })

        return true
    }


    func syncMoreContent() -> Bool {
        if isSyncing {
            return false
        }

        isSyncing = true
        isLoadingMore = true

        delegate?.syncHelper(self, syncMoreWithSuccess: {
            [weak self] (hasMore: Bool) -> Void in
            if let weakSelf = self {
                weakSelf.hasMoreContent = hasMore
                weakSelf.syncContentEnded()
            }
        }, failure: {
            [weak self] (error: NSError) -> Void in
            if let weakSelf = self {
                weakSelf.syncContentEnded()
            }
        })

        return true
    }


    // MARK: - Private Methods

    private func syncContentEnded() {
        isSyncing = false
        isLoadingMore = false

        delegate?.syncContentEnded?()
    }

}
